### ISSUE TITLE
Custom ErrorHandlers are not used by ErrorHandlerMiddleware

### ISSUE THREAD

**Issue by chinpei215** [cakephp/cakephp#11323]
This is a (multiple allowed):

* [ ] bug
* [ ] enhancement
* [x] feature-discussion (RFC)

* CakePHP Version: 3.5.3
* Platform and Target: PHP7

### What you did

Declare the following controller:
```php
class TestsController extends AppController
{
    public function exception() {
        throw new \Exception();
    }

    public function error() {
        throw new \Error();
    }
}
```
Access `/tests/exception` and `/tests/error`.

### What happened

While Exceptions are handled by `ErrorHandlerMiddleware::handleException()`, PHP7 Errors are handled by `ErrorHandler::handleException()`.

### What you expected to happen

Both Exceptions and PHP7 Errors are handled by `ErrorHandlerMiddleware::handleException()` if `ErrorHandlerMiddleware` is enabled.

### Related code

https://github.com/cakephp/cakephp/blob/35f70ad2c9843b1c42afba072615ce9141d3bf45/src/Error/Middleware/ErrorHandlerMiddleware.php#L92-L99

### Related issues

#8997, #9500, #9931

**Comment by robertpustulka** [comment]
Would something like that work both in PHP 5 and 7?

```php
         try { 
             return $next($request, $response); 
         } catch (Exception $e) { 
             return $this->handleException($e, $request, $response); 
         } catch (Throwable $e) { 
             return $this->handleException($e, $request, $response); 
         }  
```

**Comment by chinpei215** [comment]
Yes. And if there are no problems, I would like to add the same catch blocks to [Connection::transactional()](https://github.com/cakephp/cakephp/blob/7c988031eb5b5ec966a85b4a0ba0d3c44f5d7259/src/Database/Connection.php#L680-L685) too. Otherwise, if people use DatabaseLog engine, the logs are not logged if PHP7 errors happen in transactions.

**Comment by markstory** [comment]
Related to #8997 and #9500

**Comment by chinpei215** [comment]
By the way, what do you think about using the `ErrorHandler` in the `ErrorHandlerMiddleware`? Right now I can see kinds of duplicate code in the two classes:
https://github.com/cakephp/cakephp/blob/668c0873986adeec2187aaf8f798d1d62306e61a/src/Error/BaseErrorHandler.php#L350
https://github.com/cakephp/cakephp/blob/35f70ad2c9843b1c42afba072615ce9141d3bf45/src/Error/Middleware/ErrorHandlerMiddleware.php#L190

But I don't think we can deprecate `BaseErrorHandler`'s one. Because, even if there were another catch block for PHP7 Errors in the `ErrorHandlerMiddleware::handleException()`, `BaseErrorHandler::handleException()` would be still called when an exception happens before or after HTTP middleware layer. And, `ConsoleErrorHandler`, derived from `BaseErrorHandler`, would continue to be used in CLI applications (if new Command class doesn't change the current error handling process).

So I think it would be better to use the `ErrorHandler` in the `ErrorHandlerMiddleware`:
```php
try { 
    return $next($request, $response); 
} catch (Exception $e) {
} catch (Throwable $e) {
}

$this->errorHandler->setRequest($request);
$this->errorHandler->setResponse($response);
return $this->errorHandler->handleException($e); 
```

The `BaseErrorHandler::setRequest()` is a new method. If the request is given, `BaseErrorHandler` will use it instead of `Router::getRequest()`. The `BaseErrorHandler::setReponse()` may also be needed. Probably, this change would allow people to create a custom error handler in the same way as before. As of CakePHP 3.3, it is bit hard to do it.

**Comment by chinpei215** [comment]
Related to #9931

**Comment by markstory** [comment]
What is harder about creating an exception handler now compared with 3.3?

**Comment by chinpei215** [comment]
As of 3.3, ErrorHandler doesn't catch any exceptions usually, as ErrorHandlerMiddleware catches and discards them earlier than ErrorHandler. So the following section of the Cookbook is no longer accurate.
https://book.cakephp.org/3.0/en/development/errors.html#creating-your-own-error-handler
In addition to create a custom error handler, people have to extend ErrorHandlerMiddleware, or remove ErrorHandlerMiddleware from HTTP middleware stack.

**Comment by markstory** [comment]
The global error handler is still used, but only for errors before the middleware is reached. I can see how the current state is not as elegant as it could be though.

**Comment by chinpei215** [comment]
Uncatchable errors, such as E_ERROR, would also be handled by the global error handler via the handleFatalError() method. Anyway, the current error handling process seems more complex than 3.3. So I suggested to make ErrorHandlerMiddleware use ErrorHandler::handleException() instead of its own exception handling code. ErrorHandler::handleException() will come to be called for any type of errors/exceptions.

**Comment by markstory** [comment]
All PHP errors go to the existing ErrorHandler and not the middleware. Using the ErrorHandler inside the middleware presents a couple challenges. 

1. The current handleException() method directly outputs whereas the middleware needs to return a response.
2. We don't yet have a way to pass an ErrorHandler into the middleware.

These are solvable issues, but they should be figured out.

**Comment by chinpei215** [comment]
I didn't realize the second one... How about doing as follows:
```php
// In bootstrap.php
$this->setErrorHandler(new ErrorHandler(Configure::read('Error')));

// In Application.php
$errorHandler = new ErrorHandlerMiddleware();
$errorHandler->setErrorHandler($this->getErrorHandler());
$middlewareQueue->add($errorHandler);
```
Though we have to also update ConsoleApplicationInterface and HttpApplicationInterface.

**Comment by markstory** [comment]
Could the ErrorHandler be a constructor argument for the middleware? We could fall back to the current behaviour if no handler is provided.

**Comment by chinpei215** [comment]
It could. The method signature of the `ErrorHandlerMiddleware::__construct()` is the following:
```php
public function __construct($exceptionRenderer = null, array $config = []);
```
How do we pass the ErrorHandler to this? There would be several choices - using the `$config` argument, adding the third argument, or constructor overloading.

**Comment by markstory** [comment]
That is unfortunate. If an error handler instance is passed in wouldn't that be better than & replace an exception renderer being passed in?

Even if we can get an `ErrorHandler` into the middleware we'll need more methods on the ErrorHandler to return a modified response. Right now the `handleException` method outputs the response directly whereas the middleware stack will want a response to be returned.

Should we consider factoring out the currently duplicated sections of code into other objects that can be used by both the middleware and global/default handler?

**Comment by chinpei215** [comment]
> If an error handler instance is passed in wouldn't that be better than & replace an exception renderer being passed in?

Sorry, I cannot parse the structure of the avoe English sentence 😞 
Did you say, "If an error handler instance is passed in you, it would be better to replace an exception renderer too."?

> Should we consider factoring out the currently duplicated sections of code into other objects that can be used by both the middleware and global/default handler?

At first, I was planning to create a new `ExceptionLogger` class for chained exception rendering. But it doesn't solve the above problem on creating custom handlers.

**Comment by markstory** [comment]
> Sorry, I cannot parse the structure of the above English sentence

That's my mistake. It was an awkward sentence. I was trying to say that the current `exceptionRenderer` argument could be overloaded to accept either an instance of `BaseErrorHandler` or `ExceptionRendererInterface`. Getting an instance of `BaseErrorHandler` would allow `ErrorHandlerMiddleware` to use it for more logic than just getting an `ExceptionRenderer` would.

An ExceptionLogger would allow us to consolidate the code that is currently duplicated and reuse it. In the middleware the logger could be provided as a config option.

**Comment by chinpei215** [comment]
Thank you for the clarification. I will try to overload `ErrorHandlerMiddleware::__construct()`. But if it is difficult, I will give up and create `ExceptionLogger` instead.

**Comment by chinpei215** [comment]
If someone interests, please do this instead of me? I will give my priority to the other tasks.

**Comment by dereuromark** [comment]
Resolved with above PR and follow up https://github.com/cakephp/cakephp/pull/11518

**Comment by chinpei215** [comment]
Probably, I should have changed the issue title. The current error handling process is bit strange.

- (Uncatchable) Fatal errors are handled by the legacy ErrorHandler
- <del>PHP7 errors were handled by the legacy ErrorHandler</del> Changed in 3.5.7
- Exceptions are handled b y the ErrorHandlerMiddleware. (But client IP are not logged.

So I wanted to change ErrorHandlerMiddleware to use the legacy ErrorHandler. It allows people to create their custom ErrorHandler in the documented way. Currently, our Cookbook lies.

**Comment by dereuromark** [comment]
Streamlining that would probably be a good move, also remove un-DRY code here.
If you have a fix better than the PR I opened, go for it!
